README.md
# CRCcalculation_Library
CRC calculation library for 8-bit, 16-bit, 24-bit, 32-bit, 40-bit and 64 bit. Supports 76 CRC types

CRC Calculation Library provides functionality to compute Cyclic Redundancy Check (CRC) value of input data 
in the size of upto 2^64 bytes (theoratically). It supports CRC types with width in multiplier of 8-bits. 
Predefined CRC types are listed below; 

- CRC-8/AUTOSAR
- CRC-8/BLUETOOTH
- CRC-8/CDMA2000
- CRC-8/DARC
- CRC-8/DVB-S2
- CRC-8/GSM-A
- CRC-8/GSM-B
- CRC-8/HITAG
- CRC-8/I-432-1
- CRC-8/I-CODE
- CRC-8/LTE
- CRC-8/MAXIM-DOW
- CRC-8/MIFARE-MAD
- CRC-8/NRSC-5
- CRC-8/OPENSAFETY
- CRC-8/ROHC
- CRC-8/SAE-J1850
- CRC-8/SMBUS			
- CRC-8/TECH-3250 
- CRC-8/WCDMA 
- CRC-16/ARC 
- CRC-16/CDMA2000 
- CRC-16/CMS 
- CRC-16/DDS-110 
- CRC-16/DECT-R 
- CRC-16/DECT-X 
- CRC-16/DNP 
- CRC-16/EN-13757 
- CRC-16/GENIBUS 
- CRC-16/GSM 
- CRC-16/IBM-3740 
- CRC-16/IBM-SDLC 
- CRC-16/ISO-IEC-14443-3-A
- CRC-16/KERMIT 
- CRC-16/LJ1200 
- CRC-16/MAXIM-DOW 
- CRC-16/MCRF4XX 
- CRC-16/MODBUS 
- CRC-16/NRSC-5 
- CRC-16/OPENSAFETY-A 
- CRC-16/OPENSAFETY-B 
- CRC-16/PROFIBUS 
- CRC-16/RIELLO 
- CRC-16/SPI-FUJITSU 
- CRC-16/T10-DIF 
- CRC-16/TELEDISK 
- CRC-16/TMS37157 
- CRC-16/UMTS 
- CRC-16/USB 
- CRC-16/XMODEM 
- CRC-24/BLE 
- CRC-24/FLEXRAY-A 
- CRC-24/FLEXRAY-B 
- CRC-24/INTERLAKEN 
- CRC-24/LTE-A 
- CRC-24/LTE-B 
- CRC-24/OPENPGP 
- CRC-24/OS-9 
- CRC-32/AIXM 
- CRC-32/AUTOSAR 
- CRC-32/BASE91-D 
- CRC-32/BZIP2 
- CRC-32/CD-ROM-EDC 
- CRC-32/CKSUM 
- CRC-32/ISCSI 
- CRC-32/ISO-HDLC 
- CRC-32/JAMCRC 
- CRC-32/MEF 
- CRC-32/MPEG-2 
- CRC-32/XFER 
- CRC-40/GSM 
- CRC-64/ECMA-182 
- CRC-64/GO-ISO 
- CRC-64/MS 
- CRC-64/WE 
- CRC-64/XZ 

A sample console application; SampleApp_TestString is provided. It calculates CRCs for famous "123456789" string 
in supported 76 types and compares the result with known value.

The header files listed below are provided to use library;

CRCcalculation_Library.h includes; (required)
- a data structure definition if any CRC parameter is needed to be accessed by an application.
- prototypes of functions in the library.

SupportedTypes.h includes list of supported CRC types. (required)

ListOfPolynomials.h includes parameters of all supported CRC types. (if necessary)

/*This file is part of CRCcalculation_Library which is a free software : you can redistribute itand /or modify it under the terms of 
the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
CRCcalculation_Library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the GNU General Public License for more details. 
You should have received a copy of the GNU General Public License along with Foobar.If not, see < https://www.gnu.org/licenses/>.
*/


ChangeLog.txt
first release V1.0.0

SupportedTypes.h
/*This file is part of CRCcalculation_Library which is a free software : you can redistribute itand /or modify it under the terms of 
the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
CRCcalculation_Library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the GNU General Public License for more details. 
You should have received a copy of the GNU General Public License along with Foobar.If not, see < https://www.gnu.org/licenses/>.
*/

#pragma once
#define CRC_8_AUTOSAR		0
#define CRC_8_BLUETOOTH		1
#define CRC_8_CDMA2000		2
#define CRC_8_DARC			3
#define CRC_8_DVB_S2		4
#define CRC_8_GSM_A			5
#define CRC_8_GSM_B			6
#define CRC_8_HITAG			7
#define CRC_8_I_432_1		8
#define CRC_8_I_CODE		9
#define CRC_8_LTE			10
#define CRC_8_MAXIM_DOW		11
#define CRC_8_MIFARE_MAD	12
#define CRC_8_NRSC_5		13
#define CRC_8_OPENSAFETY	14
#define CRC_8_ROHC			15
#define CRC_8_SAE_J1850		16
#define CRC_8_SMBUS			17		
#define CRC_8_TECH_3250		18
#define CRC_8_WCDMA			19
#define CRC_16_ARC			20
#define CRC_16_CDMA2000		21
#define CRC_16_CMS			22
#define CRC_16_DDS_110		23
#define CRC_16_DECT_R		24
#define CRC_16_DECT_X		25
#define CRC_16_DNP			26
#define CRC_16_EN_13757		27
#define CRC_16_GENIBUS		28
#define CRC_16_GSM			29
#define CRC_16_IBM_3740		30
#define CRC_16_IBM_SDLC		31
#define CRC_16_ISO_IEC_14443_3_A	32
#define CRC_16_KERMIT		33
#define CRC_16_LJ1200		34
#define CRC_16_MAXIM_DOW	35
#define CRC_16_MCRF4XX		36
#define CRC_16_MODBUS		37
#define CRC_16_NRSC_5		38
#define CRC_16_OPENSAFETY_A 39
#define CRC_16_OPENSAFETY_B 40
#define CRC_16_PROFIBUS		41
#define CRC_16_RIELLO		42
#define CRC_16_SPI_FUJITSU	43
#define CRC_16_T10_DIF		44
#define CRC_16_TELEDISK		45
#define CRC_16_TMS37157		46
#define CRC_16_UMTS			47
#define CRC_16_USB			48
#define CRC_16_XMODEM		49
#define CRC_24_BLE			50
#define CRC_24_FLEXRAY_A	51
#define CRC_24_FLEXRAY_B	52
#define CRC_24_INTERLAKEN	53
#define CRC_24_LTE_A		54
#define CRC_24_LTE_B		55
#define CRC_24_OPENPGP		56
#define CRC_24_OS_9			57
#define CRC_32_AIXM			58
#define CRC_32_AUTOSAR		59
#define CRC_32_BASE91_D		60
#define CRC_32_BZIP2		61
#define CRC_32_CD_ROM_EDC	62
#define CRC_32_CKSUM		63
#define CRC_32_ISCSI		64
#define CRC_32_ISO_HDLC		65
#define CRC_32_JAMCRC		66
#define CRC_32_MEF			67
#define CRC_32_MPEG_2		68
#define CRC_32_XFER			69
#define CRC_40_GSM			70
#define CRC_64_ECMA_182		71
#define CRC_64_GO_ISO		72
#define CRC_64_MS			73
#define CRC_64_WE			74
#define CRC_64_XZ			75


CRCcalculation_Library.h
/*This file is part of CRCcalculation_Library which is a free software : you can redistribute itand /or modify it under the terms of 
the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
CRCcalculation_Library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the GNU General Public License for more details; https://www.gnu.org/licenses/

https://github.com/tanel312  www.tanels.com  tanel.utilities@gmail.com
*/
#pragma once

// structure of CRC parameters
typedef struct _crcparam
{
    uint8_t width;
    uint64_t normal;
    uint64_t initial;
    bool initialrefl;
    bool reflected;
    uint64_t xorout;
    uint64_t check;
    uint64_t residue;
    const char name[26];
} CRCparameters;

// Reflects the input(mirroring)
void Reflect(
    uint64_t* Inp, // input/output; value to be reflected
    uint8_t Width); // input; width of value in bits

// Creates CRC Lookup Table
void CreateCRCtable(
    uint64_t* CrcTable, // output; CRC lookup table, array of 256 cells
    uint64_t Polynomial, // input; normal polnomial
    uint8_t Width); // width of CRC in bits

// Creates CRC Reflected Lookup Table
void CreateCRCtableReflected(
    uint64_t* CrcTable, // output; CRC Reflected lookup table, array of 256 cells
    uint64_t Polynomial, // input; normal polnomial (not reflected)
    uint8_t Width); //  input; width of CRC in bits

// Calculates the CRC iaw selected type
uint64_t CalculateCRC(
    uint8_t* Buffer, // input; input data
    uint64_t Length, // input; length of input data in bytes
    uint64_t* CrcTable, // input; CRC Lookup table
    uint8_t Width, // input; width of CRC in bits
    uint64_t Initialcrc, // input; inital XOR value
    bool Initialrefl, // input; whether inital XOR value to be reflected
    uint64_t XORout); // input; final XOR value

// Calculates the reflected CRC iaw selected type
uint64_t CalculateCRCreflected(
    uint8_t* Buffer, // input; input data
    uint64_t Length, // input; length of input data in bytes
    uint64_t* CrcTable, // input; CRC Reflected Lookup table
    uint8_t Width, // input; width of CRC in bits
    uint64_t Initialcrc, // input; inital XOR value
    bool Initialrefl, // input; whether inital XOR value to be reflected
    uint64_t XORout); // input; final XOR value

// Main function to be called to calculate the CRC of input data iaw the selected type. CRC returned as unsigned long long integer
uint64_t CRC_Computation(
    uint8_t* Input, // input; data array of byte
    uint64_t Length, // input; length of input buffer in bytes
    uint8_t Type); // input; type of CRC, supported types listed in SupportedTypes.h



SampleApp_TestString/SampleApp_TestString.cpp
/* SampleApp_TestString.cpp : This file contains the 'main' function. Program execution begins and ends there.
This file is part of CRCcalculation_Library which is a free software : you can redistribute itand /or modify it under the terms of
the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
CRCcalculation_Library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the GNU General Public License for more details; https://www.gnu.org/licenses/

https://github.com/tanel312  www.tanels.com  tanel.utilities@gmail.com
*/

#include <iostream>
#include "..\CRCcalculation_Library.h"
#include "..\SupportedTypes.h"

#pragma comment( lib, "C:\\Users\\Tanel\\source\\repos\\CRCcalculation_Library\\Debug\\CRCcalculation_Library.lib" )

extern CRCparameters CRCparams[76];

int main()
{
    unsigned int count = sizeof(CRCparams) / sizeof(CRCparameters);
    uint8_t buf[] = { 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39 };
    for (unsigned int i = 0; i < count; i++)
    {
        std::cout << CRCparams[i].name;
        if (CRC_Computation(buf, sizeof(buf) / sizeof(uint8_t), i) == CRCparams[i].check)
            std::cout << " ; *** OK";
        else
            std::cout << " ; !!! NOT MATCH";
        std::cout << "\n";
    }
}


CRCcalculation_Library.cpp
/* CRCcalculation_Library.cpp : Defines the functions for the static library.

This file is part of CRCcalculation_Library which is a free software : you can redistribute itand /or modify it under the terms of 
the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.
CRCcalculation_Library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty 

https://github.com/tanel312  www.tanels.com  tanel.utilities@gmail.com
*/

#include "pch.h"
#include "framework.h"

#include "CRCcalculation_Library.h"
#include "ListOfPolynomials.h"

// Mode can be one of followings;
#define CRC8 0
#define CRC16 1
#define CRC24 2
#define CRC32 3
#define CRC40 4
#define CRC64 7

uint64_t const Mask[8] = { 0xFF, 0xFFFF, 0xFFFFFF, 0xFFFFFFFF, 0xFFFFFFFFFF, 0, 0, 0xFFFFFFFFFFFFFFFF };
uint64_t const Msbcheck[8] = { 0x80, 0x8000, 0x800000, 0x80000000, 0x8000000000, 0, 0, 0x8000000000000000 };
uint8_t const Shifter[8] = { 0, 8, 16, 24, 32, 0, 0, 56 };

/*-------------------------------------------------------------------------------
  Reflect
  Reflects the input (mirroring)
  Inputs: Width
  Input/Output: Inp
-------------------------------------------------------------------------------*/
void Reflect(uint64_t* Inp, uint8_t Width)
{
    uint64_t tmp = ((uint64_t)Width / 8) - 1;
    uint64_t mask = Msbcheck[tmp];
    tmp = *Inp;
    *Inp = 0;
    while (Width--)
    {
        if (tmp & 1L)
            *Inp |= mask;
        tmp >>= 1;
        mask >>= 1;
    }
    return;
}
/*-------------------------------------------------------------------------------
  CreateCRCtable
  Creates CRC Lookup Table
  Inputs: Polynomial, Width
  Output: CrcTable
-------------------------------------------------------------------------------*/
void CreateCRCtable(uint64_t* CrcTable, uint64_t Polynomial, uint8_t Width)
{
    uint64_t index;
    uint64_t value;
    uint8_t cnt;
    uint8_t mode = Width / 8 - 1;

    for (index = 0; index < 256; index++)
    {
        value = ((index << Shifter[mode]) & Mask[mode]);
        cnt = 8;
        while (cnt--)
        {
            value = ((value & Msbcheck[mode]) == 0) ? (value << 1) : (value << 1) ^ Polynomial;
            value &= Mask[mode];
        }
        CrcTable[index] = value;
    }
}
/*-------------------------------------------------------------------------------
  CreateCRCtableReflected
  Creates CRC Reflected Lookup Table
  Inputs: Polynomial, Width
  Output: CrcTable
-------------------------------------------------------------------------------*/
void CreateCRCtableReflected(uint64_t* CrcTable, uint64_t Polynomial, uint8_t Width)
{
    uint64_t index;
    uint64_t value, tmp;
    uint8_t cnt;
    uint8_t mode = Width / 8 - 1;
    uint64_t polynomalR = Polynomial;
    Reflect(&polynomalR, Width);
    for (index = 0; index < 256; index++)
    {
        value = 0;
        tmp = index;
        cnt = 8;
        while (cnt--)
        {
            value = (((value ^ tmp) & 0x1) == 0) ? (value >> 1) : (value >> 1) ^ polynomalR;
            value &= Mask[mode];
            tmp >>= 1;
        }
        CrcTable[index] = value;
    }
    return;
}
/*-------------------------------------------------------------------------------
  CalculateCRC
  Calculates the CRC iaw selected type
  Inputs: Buffer, Length, CrcTable, Width, Initialcrc, Initialrefl, XORout
  Output: function returns
-------------------------------------------------------------------------------*/
uint64_t CalculateCRC(uint8_t* Buffer, uint64_t Length, uint64_t* CrcTable, uint8_t Width, uint64_t Initialcrc, bool Initialrefl, uint64_t XORout)
{
    uint64_t crc = Initialcrc;
    if (Initialrefl)
        Reflect(&crc, Width);
    uint8_t mode = Width / 8 - 1;
    uint8_t index;
    while (Length--)
    {
        index = (uint8_t)((crc >> Shifter[mode]) ^ *Buffer++);
        crc = (mode == CRC8) ? (CrcTable[index]) : ((crc << 8) ^ CrcTable[index]);
    }
    return ((crc ^ XORout) & Mask[mode]);
}
/*-------------------------------------------------------------------------------
  CalculateCRCreflected
  Calculates the reflected CRC iaw selected type
  Inputs: Buffer, Length, CrcTable, Width, Initialcrc, Initialrefl, XORout
  Output: function returns
-------------------------------------------------------------------------------*/
uint64_t CalculateCRCreflected(uint8_t* Buffer, uint64_t Length, uint64_t* CrcTable, uint8_t Width, uint64_t Initialcrc, bool Initialrefl, uint64_t XORout)
{
    uint64_t crc = Initialcrc;
    if (Initialrefl)
        Reflect(&crc, Width);
    uint8_t mode = Width / 8 - 1;
    while (Length--)
    {
        crc = (crc >> 8) ^ CrcTable[(crc ^ *Buffer++) & Mask[CRC8]];
    }
    return((crc ^ XORout) & Mask[mode]);
}
/*-------------------------------------------------------------------------------
  CRC_Computation
  Main function of CRC calculation library
  Inputs: Input, Length, Type
  Output: function
-------------------------------------------------------------------------------*/
uint64_t CRC_Computation(uint8_t* Input, uint64_t Length, uint8_t Type)
{
    uint64_t crc;
    uint64_t crctable[256];

    if (CRCparams[Type].reflected == false)
    {
        // normal crc 
        CreateCRCtable(crctable, CRCparams[Type].normal, CRCparams[Type].width);
        crc = CalculateCRC(Input, Length, crctable, CRCparams[Type].width, CRCparams[Type].initial, CRCparams[Type].initialrefl, CRCparams[Type].xorout);
    }
    else
    {
        // reflected crc
        CreateCRCtableReflected(crctable, CRCparams[Type].normal, CRCparams[Type].width);
        crc = CalculateCRCreflected(Input, Length, crctable, CRCparams[Type].width, CRCparams[Type].initial, CRCparams[Type].initialrefl, CRCparams[Type].xorout);
    }
    return crc;
}
